File: DP/book-shop.cpp
#include <iostream>
#include <vector>
#include <algorithm>

int main() {
    int n, x;
    std::cin >> n >> x;

    std::vector<int> price(n);
    std::vector<int> pages(n);

    for (int i = 0; i < n; ++i) {
        std::cin >> price[i];
    }

    for (int i = 0; i < n; ++i) {
        std::cin >> pages[i];
    }

    std::vector<std::vector<int>> dp(n + 1, std::vector<int>(x + 1, 0));

    for (int i = 1; i <= n; ++i) {
        for (int j = 1; j <= x; ++j) {
            dp[i][j] = dp[i - 1][j];
            if (j - price[i - 1] >= 0) {
                dp[i][j] = std::max(dp[i][j], pages[i - 1] + dp[i - 1][j - price[i - 1]]);
            }
        }
    }

    std::cout << dp[n][x] << std::endl;

    return 0;
}

File: DP/book-shop.py
n, x = map(int, input().split())
price = [int(y) for y in input().split()]
pages = [int(z) for z in input().split()]

dp = [[0 for _ in range(x+1)] for _ in range(n+1)]

for i in range(1,n+1):
    for j in range(1,x+1):
        dp[i][j] = dp[i-1][j]
        if j-price[i-1] >= 0:
            dp[i][j] = max(dp[i][j], pages[i-1] + dp[i-1][j-price[i-1]])

print(dp[n][x])